File: node/outboxify-core/src/model/outbox-record.js
/*
 * Copyright 2026 Outboxify Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const crypto = require('crypto');
const { OutboxStatus } = require('./outbox-status');

class DefaultOutboxRecord {
  /**
   * @param {Object} options
   * @param {string} [options.outboxPipeline]
   * @param {string} [options.outboxId]
   * @param {string} options.topic
   * @param {string} [options.partitionKey]
   * @param {string} options.payload
   * @param {Record<string, string>} [options.headers]
   * @param {string} [options.status]
   * @param {number} [options.retryCount]
   * @param {string} [options.lastError]
   * @param {Date} [options.createdAt]
   * @param {Date} [options.updatedAt]
   * @param {Date} [options.processedAt]
   */
  constructor(options) {
    this.outboxPipeline = options.outboxPipeline || 'default';
    this.outboxId = options.outboxId || crypto.randomUUID();
    this.topic = options.topic;
    this.partitionKey = options.partitionKey || null;
    this.payload = options.payload;
    this.headers = options.headers || {};
    this.status = options.status || OutboxStatus.NEW;
    this.retryCount = options.retryCount || 0;
    this.lastError = options.lastError || null;
    this.createdAt = options.createdAt || new Date();
    this.updatedAt = options.updatedAt || new Date();
    this.processedAt = options.processedAt || null;
  }

  getOutboxPipeline() { return this.outboxPipeline; }
  getOutboxId() { return this.outboxId; }
  getTopic() { return this.topic; }
  getPartitionKey() { return this.partitionKey; }
  getPayload() { return this.payload; }
  getHeaders() { return this.headers; }
  getStatus() { return this.status; }
  getRetryCount() { return this.retryCount; }
  getLastError() { return this.lastError; }
  getCreatedAt() { return this.createdAt; }
  getUpdatedAt() { return this.updatedAt; }
  getProcessedAt() { return this.processedAt; }

  static fromPayload(pipeline, id, payload) {
    return new DefaultOutboxRecord({
      outboxPipeline: pipeline,
      outboxId: id,
      topic: payload.topic,
      partitionKey: payload.partitionKey,
      payload: payload.payload,
      headers: payload.headers,
      status: OutboxStatus.NEW
    });
  }
}

module.exports = {
  DefaultOutboxRecord
};

File: java/outboxify-core/src/main/java/io/outboxify/core/model/OutboxRecord.java
/*
 * Copyright 2026 Outboxify Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outboxify.core.model;

import java.time.Instant;
import java.util.Collections;
import java.util.Map;

/**
 * Universal contract representing a transactional outbox record.
 * Can be implemented directly by domain entities (e.g. JPA @Entity Order)
 * or by dedicated outbox table models.
 */
public interface OutboxRecord {

    /**
     * The pipeline identifier this record belongs to.
     * Defaults to "default" if omitted.
     *
     * @return pipeline name
     */
    default String getOutboxPipeline() {
        return "default";
    }

    /**
     * Unique identifier for this outbox row (e.g. UUID or sequence ID).
     *
     * @return unique record ID
     */
    String getOutboxId();

    /**
     * Destination message broker topic (e.g. "orders.v1").
     *
     * @return target topic
     */
    String getTopic();

    /**
     * Message partition key for ordering guarantees in Kafka / partitioned brokers.
     *
     * @return partition key, or null if unkeyed
     */
    default String getPartitionKey() {
        return null;
    }

    /**
     * Serialized message payload (JSON, Avro string, or raw text).
     *
     * @return string payload
     */
    String getPayload();

    /**
     * Metadata headers dispatched with the message (tracing, event type, correlation ID).
     *
     * @return map of string key-value header pairs
     */
    default Map<String, String> getHeaders() {
        return Collections.emptyMap();
    }

    /**
     * Current lifecycle status of the record.
     *
     * @return outbox status
     */
    default OutboxStatus getStatus() {
        return OutboxStatus.NEW;
    }

    /**
     * Number of failed delivery attempts.
     *
     * @return retry count
     */
    default int getRetryCount() {
        return 0;
    }

    /**
     * Error message or stack trace summary from the last failed attempt.
     *
     * @return last error string, or null
     */
    default String getLastError() {
        return null;
    }

    /**
     * Timestamp when the record was created.
     *
     * @return creation instant
     */
    default Instant getCreatedAt() {
        return Instant.now();
    }

    /**
     * Timestamp when the record was last updated or locked.
     *
     * @return update instant
     */
    default Instant getUpdatedAt() {
        return Instant.now();
    }

    /**
     * Timestamp when the record was confirmed SENT by the message broker.
     *
     * @return processed instant, or null if pending
     */
    default Instant getProcessedAt() {
        return null;
    }
}
